src/pages/memes/[id].tsx

```diff
@@ -1,31 +1,69 @@
-import type { NextPage } from "next";
-import { useRouter } from "next/router";
+import { dehydrate, QueryClient } from "@tanstack/react-query";
+import type { GetStaticPaths, GetStaticProps, NextPage } from "next";
 import { Suspense } from "react";
 
-import { DEFAULT_DESCRIPTION, TITLE } from "@/application/util";
+import { fetchMemeDetailById } from "@/application/hooks";
+import { TITLE } from "@/application/util";
 import { ExplorePageNavigation } from "@/components/common/Navigation";
 import { NextSeo } from "@/components/common/NextSeo";
+import { SSRSuspense } from "@/components/common/Suspense";
 import { MemeCTAList, MemeDetail, MemeTagList, RelativeMemeList } from "@/components/meme/MemeInfo";
+import type { DefaultPageProps, Meme } from "@/types";
 
-const MemeDetailPage: NextPage = () => {
-  const { query } = useRouter();
+interface Props {
+  id: string;
+  meme: Pick<Meme, "name" | "description">;
+}
 
+const MemeDetailPage: NextPage<Props> = ({ id, meme: { name, description } }) => {
   return (
     <>
-      <NextSeo description={DEFAULT_DESCRIPTION} title={TITLE.memeDetail} />
+      <NextSeo description={description} title={TITLE.memeDetail(name)} />
       <ExplorePageNavigation />
-      {query.id && (
-        <Suspense>
-          <MemeDetail id={query.id as string} />
-          <MemeTagList id={query.id as string} />
-          <MemeCTAList id={query.id as string} />
-          <Suspense>
-            <RelativeMemeList />
-          </Suspense>
-        </Suspense>
-      )}
+      <Suspense>
+        <MemeDetail id={id} />
+        <MemeTagList id={id} />
+        <MemeCTAList id={id} />
+      </Suspense>
```

Comment:
이 부분에서 정말로 server side rendering 이 되는건지 궁금합니다~

`getStaticProps` 에서 `QUERY_KEYS.getMemeDetailById(id)`라는 query key로 fetch 하고 그 QueryClient를 dehydrate 하잖아요. (이 dehydrate 한 query는 클라이언트에서 hydrate할 때 사용할 데이터겠죠.)

그 다음 server에서 rednering 할 때 cache 된 data 가 있으면 그 데이터를 가지고  render 을 진행할거고 없으면 안하잖아요. 근데 여기서 보면`QUERY_KEYS.getMemeDetailById(id)` 에 대한 query "만" fetch되어 있고 `QUERY_KEYS.getMemeTagsById(id)` 에 대한 query 는 fetch 되어 있지 않아서 서버에서 렌더링 할 때 cache 가 없으니까 렌더링이 안되는거 아닌가요??

질문의 오류가 있으면 말씀해주세용

Reply:
그리고 밈 상세 조회(`QUERY_KEYS.getMemeDetailById(id)`)API에서 에러가 나면 밈별 태그 조회(`QUERY_KEYS.getMemeTagsById(id)`)에서도 에러가 날거로 예상되는데 이 부분은 어떻게 생각하시나요?? 왜냐하면 두 API 모두 특정 밈에 대한 정보인데 오류가 나면 둘다 오류가 나지 않을까 생각합니당

Reply:
1.`getMemeTagsById`는 제가 아예 잊어먹고 있었네요..ㅎㅎ getStaticProps에 추가해놓겠습니다!

|html|브라우저에서 한번 더 요청|
|:--:|:--:|
|<img width="246" alt="image" src="https://user-images.githubusercontent.com/74011724/218241509-d40c9007-38fd-419d-91a1-e79a80442537.png">|<img width="294" alt="image" src="https://user-images.githubusercontent.com/74011724/218241553-0e5eeedf-4cbd-47ae-8353-509a1ff7bf0b.png">|
<img width="133" alt="image" src="https://user-images.githubusercontent.com/74011724/218241630-51d5f619-edf3-45a5-9dc7-cdc8597af633.png">

2. 저번에도 비슷한 얘기를 했듯이 react query의 요청이 서버 사이드에서도 일어나긴 합니다.
컴포넌트가 서버 사이드에서 프리렌더링 될 때 캐시에 데이터가 있든 없든 일단 api 요청을 하는데, getMemeTagsById를 제가 dehydrate 시키는걸 까먹었어도 **결국 api  요청 완료 후 html로 내려옵니다!** (그래서 html 내려오는 시간이 좀 늦었나..?)

참고로 dehydrate 돼서 내려와도, 클리아언트 사이드에서 다시 한번 api 요청을 합니다
- html 에선 조회수가 469인데 추가 요청에선 481 임
- 조희수 ui 가 깜빡거리긴 하지만, **최신 데이터인 481로 업데이트 됨**

3. 에러 처리는 두 api 모두 `try` 안에 넣어놓고 하나라도 오류나면 not found 내릴까 싶은데 어떨까요?

Reply:
> getMemeTagsById를 제가 dehydrate 시키는걸 까먹었어도 결국 api 요청 완료 후 html로 내려옵니다! 

이 부분이 정확히 이해가 안가는데 문서를 참조해줄 수 있을까요??
dehydrate 의 역할이 [해당 query가 나중에 hydrate될 데이터(cache)라고 해주는 거구요](https://tanstack.com/query/v4/docs/react/reference/hydration#dehydrate).(물론 stale 하게 되면 다시 query 를 요청하겠지만..)

`getMemeTagsById` query 가 prefetch 되지 않았는데(getStaticProps 에는 밈 상세 API만 fetch 하는데) 어떻게 pre rendering 될 수 있는건가요?

**useQuery hook 이 애초에 prefetch 기능이 있는건가요??**

> 참고로 dehydrate 돼서 내려와도, [클리이언트 사이드에서 다시 한번 api 요청](https://tanstack.com/query/latest/docs/react/guides/ssr?from=reactQueryV3&original=https%3A%2F%2Ftanstack.com%2Fquery%2Fv3%2Fdocs%2Fguides%2Fssr#using-hydration)을 합니다

넵 문서에서도 그렇게 말하고 있군요! prefetch한 데이터가 stale 하다면 client에서 refetch 한다구요.
> This includes refetching those queries on the client if they have become stale since the time they were rendered on the server.

---

> 3. 에러 처리는 두 api 모두 try 안에 넣어놓고 하나라도 오류나면 not found 내릴까 싶은데 어떨까요?

넵 좋습니다~

Reply:
일단 제 생각은 서버사이드에서 프리렌더링 될 때, 서스펜스 모드 `useQuery`는 api 호출을 하는 것이 맞습니다
- `getStaticProps` dehydrate에서 캐시 생성 여부와 상관 없이, **일단 컴포넌트 프리렌더링 시 useQuery api 호출**

따라서 prefetch 되지 않은 Tags api도, 프리렌더링 될 때 (캐시와 상관 없이) 어쨌든 api가 호출되기에 태그들도 html로 렌더링 되어서 내려옵니다
그러나 이 경우, cache가 생성 되어있지 않기 때문에 **브라우저 단에서 한번 더 Tags api 요청이 중복**으로 일어나게 됩니다
- 서버에서 프리렌더링 시 호출 1번, 브라우저에서 dehydrate된 캐시가 없으므로 1번

예전에 서스펜스로 삽질 할 때 알게된 거라서,, 관련된 이슈로 [next 서버에서 useQuery api 요청이 완료될 때까지 response가 blocking](https://github.com/TanStack/query/discussions/3546) 된다는 것은 있었어요 (정확히 원인이 설명된 문서는 그때도 찾기가 어려웠네요 ㅠㅠ)
- [이슈 만드신 분이 작성한 블로그 글](https://seonghyeonkimm.github.io/blog/nextjs-develop-and-operation-2)

Reply:
일단은 useQuery 가 서버에서도 실행이 된다고 받아들일게용
suspense 모드일 때랑 아닐때랑 useQuery 가 어떻게 동작하는지는 문서 읽어보고 보내주신 자료도 보면서 이해해볼게용..ㅎㅎ
긴가민가한 suspense 요놈..